pkg/alertmanager/state_replication.go

```diff
@@ -0,0 +1,212 @@
+func newReplicatedStates(userID string, rf int, f func(context.Context, string, *clusterpb.Part) error, stopc chan struct{}, l log.Logger, r prometheus.Registerer) *state {
+	partialStateMergesTotal := promauto.With(r).NewCounterVec(prometheus.CounterOpts{
+		Name: "alertmanager_partial_state_merges_total",
+		Help: "Number of times we have received a partial state to merge for a key.",
+	}, []string{"key"})
+	partialStateMergesFailed := promauto.With(r).NewCounterVec(prometheus.CounterOpts{
+		Name: "alertmanager_partial_state_merges_failed_total",
+		Help: "Number of times we have failed to merge a partial state received for a key.",
+	}, []string{"key"})
+	stateReplicationTotal := promauto.With(r).NewCounterVec(prometheus.CounterOpts{
+		Name: "alertmanager_state_replication_total",
+		Help: "Number of times we have tried to replicate a state to other alertmanagers.",
+	}, []string{"key"})
+	stateReplicationFailed := promauto.With(r).NewCounterVec(prometheus.CounterOpts{
+		Name: "alertmanager_state_replication_failed_total",
+		Help: "Number of times we have failed to replicate a state to other alertmanagers.",
+	}, []string{"key"})
+
+	s := &state{
+		logger:                   l,
+		userID:                   userID,
+		replicateStateFunc:       f,
+		replicationFactor:        rf,
+		states:                   make(map[string]cluster.State, 2), // we use two, one for the notifications and one for silences.
+		msgc:                     make(chan *clusterpb.Part),
+		readyc:                   make(chan struct{}),
+		stopc:                    stopc,
+		reg:                      r,
+		partialStateMergesTotal:  partialStateMergesTotal,
+		partialStateMergesFailed: partialStateMergesFailed,
+		stateReplicationTotal:    stateReplicationTotal,
+		stateReplicationFailed:   stateReplicationFailed,
+	}
+
+	go s.handleBroadcasts()
+	return s
+}
+
+// AddState adds a new state that will be replicated using the ReplicationFunc. It returns a channel to which the client can broadcast messages of the state to be sent.
+func (s *state) AddState(key string, cs cluster.State) *stateChannel {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	s.states[key] = cs
+
+	s.partialStateMergesTotal.WithLabelValues(key)
+	s.partialStateMergesFailed.WithLabelValues(key)
+	s.stateReplicationTotal.WithLabelValues(key)
+	s.stateReplicationFailed.WithLabelValues(key)
+
+	send := func(b []byte) {
+		s.msgc <- &clusterpb.Part{Key: key, Data: b}
+	}
+
+	return newStateChannel(key, send, s.logger, s.reg)
+}
+
+// MergePartialState merges a received partial message with an internal state.
+func (s *state) MergePartialState(p *clusterpb.Part) error {
+	s.partialStateMergesTotal.WithLabelValues(p.Key).Inc()
+
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	st, ok := s.states[p.Key]
+	if !ok {
+		level.Debug(s.logger).Log("msg", "state key not found while merging", "key", p.Key)
+		s.partialStateMergesFailed.WithLabelValues(p.Key).Inc()
+		return nil
```

Comment:
Why not returning error?

Reply:
Good catch, I was thinking I wouldn't return any merge errors given these messages are supposed to come over UDP (we settle the gossip before sending an alert) but I think it's worth saying something here.